### ISSUE TITLE
[Feat] Resume the downloads/uploads sessions

### ISSUE THREAD

**Issue by art435** [iyear/tdl#86]
When the users dealt with numerous files that need to be downloaded or uploaded, a handy feature is to be able to resume next day from where it has left.

As an example is someone tries to back up an entire channel, this could take days, and needs to be done across a few sessions.  The program should keep an index of the last file_id downloaded and continue from that the next session.

 When uploading, the program should keep an index of all the files from a folder, remember the last file uploaded and continue from that the next session.

**Comment by iyear** [comment]
Is `--skip-same` an alternative?

**Comment by art435** [comment]
I exported chat history (about 90 files), then proceed to download the files. After, I uploaded 35 more files to the chat, exported chat history again, and then I tried to download with `--skip-same`.

It took about 95 seconds (the time needed to compare files) before starting to download the new 35 files. Scaling to thousands of files results in a disproportionate waiting time.

I guess a better solution is needed.

**Comment by iyear** [comment]
When you re-export, is it possible to specify a new msg id to avoid exporting duplicate files?

Adding a `skip N` option seems difficult because the user can combine multiple sources and skipping a particular source N messages makes the parameters complex.

**Comment by iyear** [comment]
For the default configuration, the download filename contains the msg id, perhaps as a temporary solution. Showing the last msg id is still a problem for multiple sources. I have no idea how to better display them.

I will file a new issue to track the exporting issue.

**Comment by SavSanta** [comment]
Thank you for the excellent program. I used it the 5 days ago to download 850 videos using a server. No problem unlike the issues I was having in telegram_media_downloader

Although, I did notice I could not turn off the VM to increase memory and pickup where I left off.

So I too thought it would be nice to have I have a dl resumable. Unfortunaely too late for my case but I did spend a few days analyzing the code and made a preliminary WIP pull request for your testing that is a rudimentary implementation for the `dl` at https://github.com/iyear/tdl/pull/90

Let me know your thoughts

**Comment by iyear** [comment]
@art435 @SavSanta First of all, thank you both very much for your contributions and feedback.

I didn't expect anyone to use `tdl` to download files on such a large scale:). It seems that resumable download is a necessary feature.
After some thought, I think introducing the concept of `session` is unnecessary. Keeping it simple is also part of `tdl`, and too many options and concepts can be confusing for users.

So I wanted to put `resumable download` implicitly in `tdl`.
The details are as follows:

- Generate a fingerprint based on multiple sources data each time when start to download. If the fingerprint already exists and is in an incomplete state, `tdl` will ask user: "continue or re-download?".
- The current progress is also recorded during the download and persisted to the kv database

So there is no additional learning cost for the user..

Any other ideas?

**Comment by SavSanta** [comment]
Hi @iyewr,

I dont fully understand your concepts (is fingerprint).

I think it will be better to discuss further technical architecture, design, and also answer my own set of questions in the actual PR  #90 as I think this will require many function signature changes and as well as concur or not concur with the small design change ive already made in the PR

**Comment by art435** [comment]
@SavSanta Thank you for this PR, I hope that @iyear chooses the best solution.

@iyear Keeping things simpler is the best. Activate the Discussion, I have some questions about technical stuff.

**Comment by iyear** [comment]
Thanks for the PR, but exposing the session to the user will confuse them even more.

fingerprint is a hash of multi sources and is used to compare sources. And then check progress.

**Comment by iyear** [comment]
@SavSanta @art435 
I'm going to start implementing this feature. How about my solution? 

Also thanks a lot for the PR and the feedback, but PR's solution seems a bit not user-friendly.

**Comment by SavSanta** [comment]
Hi @iyear,

1. I didnt understand what you wrote or your intended solution over the last 3 weeks. You have not provided detail.


2. *More important** I asked you to answer my questions in the PR. NONE of my DESIGN related questions in the PR have been answered (look at the PR not in Issues). I said to move the discussion to the PR. Answer my questions, so that we could have a formal and common design document understand what needs to be changed.


So no sir. I really dont know what you want to do or how you wanted to implement this 🙍...

I think to continue we must **schedule a call or videochat and talk.**. 

Reccommemd a **date, time, and timezone** on Saturday or Sunday. Then maybe we can make a Telegram call in a private room or somewhere else.
 
Thanks and best regards

**Comment by iyear** [comment]
@SavSanta 
Thanks! I've read your PR weeks ago. The main difference between my idea and yours is to hide the `.tgresume` file and save the breakpoint data in the existing boltdb storage.

And try to restate my solution: for a particular multi sources, generate a hash(fingerprint) for it. We save the download progress of this fingerprint in boltdb, and each time we start a download, we can use the fingerprint to determine if the download is currently ready to continue and ask the user.

I speak very poor English, so I may not understand what you are talking about😂

**Comment by SavSanta** [comment]
@iyear

I somewhat may understand now one of your changes at least to "tgresume"

If you are awake now. Let us try to talk https://t.me/+W_-g6G1rV81lNzMx

**Comment by iyear** [comment]
How about discussing it in a GitHub Issue? :)

It will help others understand the history of the design

**Comment by SavSanta** [comment]
No. I do not think.

because i still havent had my questions answered in the PR. 

You havent presented one technical document..

In either regards. Ive tried to assist but i wont be able to help further. Good luck!

**Comment by iyear** [comment]
emmm... .ok, thanks again. I will consider implementing it

**Comment by art435** [comment]
I think it is better to have the discussion here, so others can see it and give their opinions. I will open a new issue that could benefit this function of mass downloading.

**Comment by SavSanta** [comment]
@art435 With respect, but if you're not writing code. It doesn't matter for you since it isn't **technical ** . And anybody with actual codeling or technical skill couldve moved to the PR.

So discussion is for side guys but the PR is for to addressing **technical** matters on existing code and implementation. Not for the side gallery of people who aren't doing any actual work. Yet iyear declined to answer and he refused to participate in for a full **technicial architecture** exploration in call. So I realize after 3 weeks its' a waste of time really for me to continue. I'm sure whatever he rolls out will hopefully work and be well fully thought out tho

**Comment by art435** [comment]
If the discussion is public, people with expertise can participate and give feedback.

**Comment by iyear** [comment]
The new version has been released, feel free to try it.
